File: py/Sorting.py
def merge(left: list, right: list):
    sorted_array = []
    left_index = 0
    right_index = 0

    while left_index < len(left) and right_index < len(right):
        if left[left_index] < right[right_index]:
            sorted_array.append(left[left_index])
            left_index += 1
        else:
            sorted_array.append(right[right_index])
            right_index += 1

    while left_index < len(left):
        sorted_array.append(left[left_index])
        left_index += 1

    while right_index < len(right):
        sorted_array.append(right[right_index])
        right_index += 1

    return sorted_array


def merge_sort(array: list):
    if len(array) == 1:
        return array

    else:
        mid_array = len(array) // 2
        left_array = array[:mid_array]
        right_array = array[mid_array:]
        sorted_left_array = merge_sort(left_array)
        sorted_right_array = merge_sort(right_array)

        return merge(sorted_left_array, sorted_right_array)


print(merge_sort([1, 3, 5, 7, 9, 2, 4, 6, 8, 0]))

File: src/Sorting.js
var baseArray = [9, 2, 5, 6, 4, 3, 7, 10, 1, 8]



/*  ------------- Merge Sort (Div&Conq) ----------------- */


function mergeSort(array) {
  if (array.length === 1) return array

  const mid = Math.floor(array.length / 2)

  const left = array.slice(0, mid)
  const right = array.slice(mid)
  const sortedLeft = mergeSort(left)
  const sortedRight = mergeSort(right)

  return merge(sortedLeft, sortedRight)
}

function merge(left, right) {
  let result = []
  let indexLeft = 0
  let indexRight = 0

  while (indexLeft < left.length && indexRight < right.length) {
    if (left[indexLeft] < right[indexRight]) {
      result.push(left[indexLeft])
      indexLeft++
    } else {
      result.push(right[indexRight])
      indexRight++
    }
  }

  return result.concat(left.slice(indexLeft)).concat(right.slice(indexRight))
}

/*  ------------- Merge Sort ----------------- */


